internal/component/loki/source/api/api_test.go

```diff
@@ -609,8 +612,16 @@ func testArgsWithPorts(httpPort int, grpcPort int) Arguments {
 	}
 }
 
-func getFreePort(t *testing.T) int {
-	port, err := freeport.GetFreePort()
-	require.NoError(t, err)
-	return port
+func getFreePort(t *testing.T, exclude ...int) int {
+	const maxRetries = 10
+	for range maxRetries {
```

Comment:
Looking at the implementation here I think this has the same race condition I've seen else where you ask for an unused port it will lock it but as soon as the function executes it unlocks it. Another test could call and lock the same port before the server which request the port could use it causing the error.

I found switching to use random ports in tests to be a way more effective solution.